tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,276 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/slice_sinker.h"
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include "absl/types/optional.h"
+#include "absl/container/flat_hash_set.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+
+namespace xla {
+
+namespace {
+
+class SliceSinkerImpl {
+ public:
+  // Returns whether the instruction has been visited. The visited instruction
+  // has been grouped in a true user group with its peers and it should not be
+  // grouped in another true user group.
+  bool IsVisited(const HloInstruction* instruction) const;
+
+  // Collects true operands of inst. The inst's operands should all be slices,
+  // and the true operands are the operands of slices. Returns true operand
+  // vector if all true operands are found, or return an empty vector.
+  absl::optional<std::vector<HloInstruction*>>
+  FindSourceOperandsOfSlicesForElementwiseOperation(const HloInstruction* inst);
+
+  // Collects the true user group included inst. The slice from true operand is
+  // operand-slice. The true user's operands are all operand-slices.
+  // The operand-slices of a true user should be sliced from the same range of
+  // the true operands. It means the operand-slices have the same slice_starts,
+  // slice_limits, and slice_strides.
+  // The true users in a group have the same opcode and the same true operands
+  // in order. Of course, the inst is a true user, and the others in the group
+  // are the inst's peers.
+  // Returns an empty vector if the cost of change is more than the profit.
+  // Additionally, records the user which has been visited in a true user group,
+  // and skips it in later traverse.
+  absl::optional<std::vector<HloInstruction*>> FindPeerElementwiseOperations(
+      const HloInstruction* inst,
+      const std::vector<HloInstruction*>& operands);
+
+ private:
+  absl::flat_hash_set<const HloInstruction*> visited_;
+};
+
+// Computes the cost of implimentation of sinking slice, and returns whether
+// it should be changed.
+bool ShouldReplace(const std::vector<HloInstruction*>& operands,
```

Comment:
This routine assumes that we will generate a new elementwise operation using the "slice_sources" as operands.  For the following example,  we only need the result of add([0:9] slice of p) while we generate add(p). Can we document why we make such a decision here?

p = f32[10] parameter(0)
a = f32[8] slice(p), slice=[0:8]
aa = add(a, a)
b = f32[7] slice(p), slice=[2:9]
bb = add(b, b)